sup3r/preprocessing/data_handling/exogenous_data_handling.py

```diff
@@ -19,6 +19,172 @@
 logger = logging.getLogger(__name__)
 
 
+class SingleExoDataStep(dict):
+    """Special dictionary class for exogenous_data step"""
+
+    def __init__(self, feature, combine_type, model, data):
+        """exogenous_data step dictionary for a given model step
+
+        Parameters
+        ----------
+        feature : str
+            Name of feature corresponding to `data`.
+        combine_type : str
+            Specifies how the exogenous_data should be used for this step. e.g.
+            "input", "layer", "output". For example, if tis equals "input" the
+            `data` will be used as input to the forward pass for the model step
+            given by `model`
+        model : int
+            Specifies the model index which will use the `data`. For example,
+            if `model` == 1 then the `data` will be used according to
+            `combine_type` in the 2nd model step in a MultiStepGan.
+        data : tf.Tensor | np.ndarray
+            The data to be used for the given model step.
+        """
+        step = {'model': model, 'combine_type': combine_type, 'data': data}
+        for k, v in step.items():
+            self.__setitem__(k, v)
+        self.feature = feature
+
+    @property
+    def shape(self):
+        """Shape of data array for this model step."""
+        return self['data'].shape
+
+
+class ExoData(dict):
+    """Special dictionary class for multiple exogenous_data steps"""
+
+    def __init__(self, steps):
+        """Combine multiple SingleExoDataStep objects
+
+        Parameters
+        ----------
+        steps : list | dict
+            List of SingleExoDataStep objects or a feature dictionary with list
+            of steps for each feature
```

Comment:
I think this is the place to be a lot more descriptive and give an example. Not immediately obvious to me how it could be either a list or a dict. Do we even need to handle both? Python is flexibly typed but that doesn't mean it's always a good idea to accept multiple types. Often much cleaner and more explicit if we just lean into a single type spec. 

Also, I suggest giving an example like this (copied from the previous docstrings): 

```
            Dictionary of exogenous feature data with entries describing
            whether features should be combined at input, a mid network layer,
            or with output. This doesn't have to include the 'model' key since
            this data is for a single step model. e.g.
            {'topography': {'steps': [
                {'combine_type': 'input', 'data': ..., 'resolution': ...},
                {'combine_type': 'layer', 'data': ..., 'resolution': ...}]}}